src/search.coffee

```diff
@@ -6,234 +6,252 @@ url = require 'url'
+
+        callback null, data
+
+  counts: (query, options..., callback) ->
+    sources = []
+
+    sources.push
+      key: 'web'
+      method: _.bind @rawWeb, this
+    sources.push
+      key: 'image'
+      method: _.bind @rawImages, this
+    sources.push
+      key: 'video'
+      method: _.bind @rawVideos, this
+    sources.push
+      key: 'news'
+      method: _.bind @rawNews, this
+
+    executeSearchForCounts = (source, callback) ->
+      ###
+      We only need to one result (zero is impossible) from most verticals. With
+      a web search, however, the `totalEstimatedMatches` needs to be checked
+      from a higher page for accurate data.
+      
+      The 1,000 value comes from empirical data. It seems that after 600
+      results, the accuracy gets quite consistent and accurate. I picked 1,000
+      just to be in the clear. It also doesn't matter if there are fewer than
+      1,000 results.
+      ###
+      pagination = {offset: 0, top: 1}
+      if source.key is 'web'
+        _.each pagination, (value, key) ->
+          pagination[key] += 1000
+
+      source.method query, _.defaults(pagination, options), callback
+
+    async.mapLimit sources, @parallelLimit, executeSearchForCounts,
+      (err, responses) ->
+        return callback err if err
 
-      callback null,
-        web: @extractWebResults convertToSingleSource results, 'Web'
-        images: @extractImageResults convertToSingleSource results, 'Image'
-        videos: @extractVideoResults convertToSingleSource results, 'Video'
-        news: @extractNewsResults convertToSingleSource results, 'News'
-        spelling: @extractSpellResults convertToSingleSource results,
-          'SpellingSuggestions'
-        related: @extractRelatedResults convertToSingleSource results,
-          'RelatedSearch'
+        data = {}
+        data[source.key] = 0 for source in sources
+
+        responses.forEach (response, i) ->
+          data[sources[i].key] = response.estimatedCount
+
+        callback null, data
+
+  # This modifies the endpoint used for searching to retrieve params specific
+  # to separate verticals (i.e. width/height for images and runtime for videos).
+  verticalSearch: (vertical, q, options, callback) ->
+    options = _.extend {}, options, {q, endpoint: "#{vertical}/search"}
+    @parallelSearch options, callback
+
+  # This sends the vertical via the responseFilters query param for methods
+  # which don't have specific verticals (i.e. web, spelling, related).
+  filteredSearch: (responseFilter, q, options, callback) ->
+    options = _.extend {}, options, {q, responseFilter}
+    @parallelSearch options, callback
+
+  rawWeb: (query, options, callback) ->
+    @filteredSearch 'webpages', query, options, callback
+
+  web: (query, options..., callback) ->
+    options = options[0] or {}
+    @rawWeb query, options, (err, data) =>
+      callback err, (@extractWebResults data unless err)
+
+  extractWebResults: ({results} = []) ->
```

Comment:
Default argument should be an empty object. This would compile to:

`results = [].results`